rest/route/sns.go

```diff
@@ -499,12 +557,64 @@ func (sns *ecsSNS) cleanupUnrecognizedPod(ctx context.Context, details ecsEventD
 
 	grip.Info(message.Fields{
 		"message":        "successfully cleaned up unrecognized pod",
-		"task":           details.TaskARN,
-		"cluster":        details.ClusterARN,
-		"last_status":    details.LastStatus,
-		"desired_status": details.DesiredStatus,
+		"task":           detail.TaskARN,
+		"cluster":        detail.ClusterARN,
+		"last_status":    detail.LastStatus,
+		"desired_status": detail.DesiredStatus,
 		"route":          "/hooks/aws/ecs",
 	})
 
 	return nil
 }
+
+// listECSTasks returns the ARNs of all ECS tasks running in the container
+// instance associated with the event.
+func (sns *ecsSNS) listECSTasks(ctx context.Context, details ecsContainerInstanceEventDetail) ([]string, error) {
+	// Spot check that the notification is actually for a pod in an
+	// Evergreen-owned cluster.
+	var isInManagedCluster bool
+	for _, c := range sns.env.Settings().Providers.AWS.Pod.ECS.Clusters {
+		if strings.Contains(details.ClusterARN, c.Name) {
+			isInManagedCluster = true
+			break
+		}
+	}
+	if !isInManagedCluster {
+		return nil, nil
+	}
+
+	c, err := sns.makeECSClient(sns.env.Settings())
+	if err != nil {
+		return nil, errors.Wrap(err, "getting ECS client")
+	}
+	defer c.Close(ctx)
+
+	var token *string
+	var taskARNs []string
+
+	for {
+		resp, err := c.ListTasks(ctx, &awsECS.ListTasksInput{
+			Cluster:           aws.String(details.ClusterARN),
+			ContainerInstance: aws.String(details.ContainerInstanceARN),
+			NextToken:         token,
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "listing ECS tasks")
+		}
+		if resp == nil {
+			return nil, errors.New("listing ECS tasks returned no response")
+		}
+
+		for _, arn := range resp.TaskArns {
+			if arn == nil {
+				continue
+			}
+			taskARNs = append(taskARNs, utility.FromStringPtr(arn))
+		}
+
+		token = resp.NextToken
```

Comment:
Nah, this is a completely reasonable question, and I did think about this when first writing this loop. The supposed behavior of the ECS API is that it will only return you a next token when there are actually more result page beyond the one you just asked for, so there's no hypothetical limit to how many times we could receive a next token. We won't actually infinitely loop here because the `context.Context` (which comes from the request) will eventually a timeout and fail the `ListTasks` request because the app server has a hard 60 second limit on request duration (to protect itself from unreasonably expensive requests).

To be extra safe, we can cap the infinite loop to a fixed number of since there shouldn't be an infinite number of containers running on a single container instance. I changed the loop to cap the max iterations and avoid the potential for infinite looping.